include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b, AMatrixType &A, PMatrixType &P);
+
+    /**
+     * Callback for the computation of the nonlinear residual $F(x)$.
+     */
+    std::function<int(const VectorType &x, VectorType &res)> residual;
+
+    /**
+     * Callback for the computation of the Jacobian
+     * $\dfrac{\partial F}{\partial x}$
+     */
+    std::function<int(const VectorType &x, AMatrixType &A, PMatrixType &P)>
+      jacobian;
+
+    /**
+     * Callback for monitoring the solution process.
+     *
+     * This function is called by NonlinearSolver at the beginning
+     * of each time step.
+     */
+    std::function<
+      int(const VectorType &x, const unsigned int step_number, real_type fval)>
+      monitor;
+
+    /**
+     * Setup Jacobian callback without matrices.
+     *
+     * This callback gives full control to users to set up the tangent
+     * operator $\dfrac{\partial F}{\partial x}$
+     *
+     * Solvers must be provided via NonlinearSolver::solve_for_jacobian_system.
+     */
+    std::function<int(const VectorType &x)> setup_jacobian;
+
+    /**
+     * Solution of the Jacobian system set up with
+     * NonlinearSolver::setup_jacobian
+     *
+     * This is used as a preconditioner inside the Krylov process.
+     */
+    std::function<int(const VectorType &src, VectorType &dst)>
+      solve_for_jacobian_system;
+
+    /**
+     * Callback for the computation of the energy function.
+     *
+     * This is usually not needed, since by default SNES assumes that the
+     * objective function to be minimized is $\frac{1}{2} || F(x) ||^2 $.
+     *
+     * However, if the nonlinear equations are derived from energy arguments, it
+     * may be useful to use this callback to perform linesearch or to test for
+     * the reduction in a trust region step.
+     */
+    std::function<int(const VectorType &x, real_type &obj)> energy;
+
+  protected:
+    /**
+     * The PETSc SNES object
+     */
+    SNES snes;
+
+    SmartPointer<AMatrixType, NonlinearSolver> A;
+    SmartPointer<PMatrixType, NonlinearSolver> P;
+    SmartPointer<VectorType, NonlinearSolver>  B;
+
+    /**
+     * This flag is used to support older versions of PETSc
+     */
+    bool need_dummy_assemble;
+  };
+
+#  ifndef DOXYGEN
+  /* Inline functions */
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  inline SNES
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::petsc_snes()
+  {
+    return snes;
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  inline MPI_Comm
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::get_mpi_communicator()
+    const
+  {
+    return PetscObjectComm(reinterpret_cast<PetscObject>(snes));
+  }
+#  endif
```

Comment:
Why are these in the header file?

Reply:
They must be here or in the `snes.templates.h` . If you prefer the latter, I will move them there

Reply:
moved